homeassistant/components/husqvarna_automower/__init__.py

```diff
@@ -0,0 +1,100 @@
+"""The Husqvarna Automower integration."""
+from asyncio.exceptions import TimeoutError as AsyncioTimeoutError
+import contextlib
+import logging
+from typing import cast
+
+import aioautomower
+
+from homeassistant.components.application_credentials import DATA_STORAGE
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_TOKEN
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AutomowerDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching Husqvarna data."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize data updater."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+        )
+        api_key = None
+        ap_storage: dict = cast(dict, hass.data.get("application_credentials"))
+        ap_storage_data: dict = ap_storage[DATA_STORAGE].__dict__["data"]
+        for k in ap_storage_data:
+            api_key = ap_storage_data[k]["client_id"]
+        entry_dict = entry.as_dict()
+        access_token = entry_dict["data"]["token"]
+        scope = entry_dict["data"]["token"]["scope"]
+        if "amc:api" not in scope:
+            async_create_issue(
+                hass,
+                DOMAIN,
+                "wrong_scope",
+                is_fixable=False,
+                severity=IssueSeverity.WARNING,
+                translation_key="wrong_scope",
+            )
+        low_energy = False
+        self.session = aioautomower.AutomowerSession(api_key, access_token, low_energy)
+        self.session.register_token_callback(
+            lambda token: hass.config_entries.async_update_entry(
+                entry,
+                data={"auth_implementation": DOMAIN, CONF_TOKEN: token},
+            )
+        )
+
+    async def _async_update_data(self) -> None:
+        """Fetch data from Husqvarna."""
+        try:
+            await self.session.connect()
+        except AsyncioTimeoutError as error:
+            _LOGGER.debug("Asyncio timeout: %s", error)
+            raise ConfigEntryNotReady from error
+        except Exception as error:
+            _LOGGER.debug("Exception in async_setup_entry: %s", error)
+            raise ConfigEntryAuthFailed from Exception
```

Comment:
Why would a general Exception cause a reauth? Wouldn't having no internet trigger this (as that raises a CannotConnect error or something along those lines) That should raise `UpdateFailed` instead.

Reply:
Yes, that's to general